includes/User.php

```diff
@@ -2082,68 +2061,60 @@ public static function clearUserCache( $userId ) {
 	 * wipes it, so the account cannot be logged in until
 	 * a new password is set, for instance via e-mail.
 	 *
-	 * @param $str String New password to set
-	 * @throws PasswordError on failure
+	 * @param string $password New password to set
+	 * @param bool   $forceLogout
 	 *
-	 * @return bool
+	 * @return bool on failure
+	 *
+	 * @throws Exception
+	 * @throws PasswordError on failure
 	 */
-	public function setPassword( $str, $forceLogout = true ) {
-		global $wgAuth;
+	public function setPassword( $password, $forceLogout = true ) {
+		$heliosClient = self::heliosClient();
+		$heliosPasswordChange = null;
 
-		if( $str !== null ) {
-			if( !$wgAuth->allowPasswordChange() ) {
-				throw new PasswordError( wfMsg( 'password-change-forbidden' ) );
-			}
-
-			if( !$this->isValidPassword( $str ) ) {
-				global $wgMinimalPasswordLength;
-				$valid = $this->getPasswordValidity( $str );
-				if ( is_array( $valid ) ) {
-					$message = array_shift( $valid );
-					$params = $valid;
-				} else {
-					$message = $valid;
-					$params = array( $wgMinimalPasswordLength );
-				}
-				throw new PasswordError( wfMsgExt( $message, array( 'parsemag' ), $params ) );
+		try {
+			$heliosPasswordChange = $heliosClient->setPassword( $this->getId(), $password );
+			if ( !$heliosPasswordChange ) {
+				throw new PasswordError( wfMessage( 'externaldberror' )->text() );
 			}
+		} catch ( \Exception $e ) {
+			WikiaLogger::instance()->error( 'Failed to communicate with Helios for password set', [
+				'userId' => $this->getId(),
+				'err'    => $e,
+			] );
 		}
 
-		if( !$wgAuth->setPassword( $this, $str ) ) {
-			throw new PasswordError( wfMsg( 'externaldberror' ) );
+		if ( !empty( $heliosPasswordChange->errors ) ) {
+			WikiaLogger::instance()->error( 'Helios password set failed', [
+				'userId' => $this->getId(),
+				'err'    => $heliosPasswordChange,
+			] );
+			throw new PasswordError( wfMessage( $heliosPasswordChange->errors[0]->description )->text() );
```

Comment:
Does it mean that in case heliosPasswordChange is empty, we're not throwing any exceptions (as the previous "throw new Password" is inside a catch block).
I general I think we should clean up the error-handling and -logging flow here.